gobblin-kafka/src/main/java/gobblin/kafka/writer/KafkaDataWriter.java

```diff
@@ -38,51 +47,91 @@
+    setDefaultIfUnset(producerProperties, VALUE_SERIALIZER_CONFIG, DEFAULT_VALUE_SERIALIZER);
+    setDefaultIfUnset(producerProperties, CLIENT_ID_CONFIG, CLIENT_ID_DEFAULT);
+    return producerProperties;
+  }
 
-  private static Properties getProducerProperties(Properties props)
+  private static void setDefaultIfUnset(Properties props, String key, String value)
   {
-    Properties producerProperties = stripPrefix(props, KafkaWriterConfigurationKeys.KAFKA_PRODUCER_CONFIG_PREFIX);
+    if (!props.containsKey(key)) {
+      props.setProperty(key, value);
+    }
+  }
 
-    //Provide default properties if not set from above
-    if (!producerProperties.containsKey(KEY_SERIALIZER_CONFIG)) {
-      producerProperties.setProperty(KEY_SERIALIZER_CONFIG, DEFAULT_KEY_SERIALIZER);
+  public static KafkaProducer getKafkaProducer(Properties props)
+  {
+    Config config = ConfigFactory.parseProperties(props);
+    String kafkaProducerClass = ConfigUtils.getString(config, KafkaWriterConfigurationKeys.KAFKA_WRITER_PRODUCER_CLASS, "");
+    Properties producerProps = getProducerProperties(props);
+    if (kafkaProducerClass.isEmpty())
+    {
+      return new KafkaProducer<>(producerProps);
     }
-    if (!producerProperties.containsKey(VALUE_SERIALIZER_CONFIG)) {
-      producerProperties.setProperty(VALUE_SERIALIZER_CONFIG, DEFAULT_VALUE_SERIALIZER);
+    else
+    {
+      try {
+        Class<?> producerClass = (Class<?>) Class.forName(kafkaProducerClass);
+        KafkaProducer producer = (KafkaProducer) ConstructorUtils.invokeConstructor(producerClass, producerProps);
+        return producer;
+      } catch (ClassNotFoundException | NoSuchMethodException | IllegalAccessException | InstantiationException | InvocationTargetException e) {
+        log.error("Failed to instantiate Kafka producer " + kafkaProducerClass + " as instance of KafkaProducer.class", e);
+        throw Throwables.propagate(e);
+      }
     }
-    return producerProperties;
   }
 
-
   public KafkaDataWriter(Properties props) {
-    this(new KafkaProducer<>(getProducerProperties(props)), ConfigFactory.parseProperties(props));
+    this(getKafkaProducer(props), ConfigFactory.parseProperties(props));
   }
 
   public KafkaDataWriter(KafkaProducer producer, Config config)
   {
+    recordsProduced = 0;
     recordsWritten = new AtomicInteger(0);
+    recordsFailed = 0;
     bytesWritten = new AtomicInteger(-1);
-
+    lock = new ReentrantLock();
+    stateChange = lock.newCondition();
     this.topic = config.getString(KafkaWriterConfigurationKeys.KAFKA_TOPIC);
+    this.commitTimeoutInNanos = ConfigUtils.getLong(config, KafkaWriterConfigurationKeys.COMMIT_TIMEOUT_MILLIS_CONFIG,
+        KafkaWriterConfigurationKeys.COMMIT_TIMEOUT_MILLIS_DEFAULT) * MILLIS_TO_NANOS;
     this.producer = producer;
     this.producerCallback = new Callback() {
       @Override
       public void onCompletion(RecordMetadata metadata, Exception exception) {
```

Comment:
Will this work with high volumes of data? Won't it slow down significantly? Especially with the lock?